dsgrid/project.py

```diff
@@ -130,3 +168,115 @@ def _iter_datasets(self):
 
     def list_datasets(self):
         return [x.dataset_id for x in self._iter_datasets()]
+
+    @track_timing(timer_stats_collector)
+    def process_query(self, context: QueryContext):
+        self._handle_supplemental_query_names(context)
+        self._build_filtered_record_ids_by_dimension_type(context)
+
+        dfs = []
+        for dataset_id in context.model.project.dataset_ids:
+            logger.info("Start processing query for dataset_id=%s", dataset_id)
+            dataset = self.get_dataset(dataset_id)
+            dfs.append(dataset.get_dataframe(context))
+            logger.info("Finished processing query for dataset_id=%s", dataset_id)
+
+        if not dfs:
+            logger.warning("No data matched %s", context.name)
+            return None
+
+        main_df = dfs[0]
+        if len(dfs) > 1:
+            for df in dfs[1:]:
+                if df.columns != main_df.columns:
+                    raise Exception(
+                        f"BUG: dataset columns must match. main={main_df.columns} new={df.columns}"
+                    )
+                main_df = main_df.union(df)
+
+        return self._convert_columns_to_query_names(main_df)
+
+    def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+        record_ids = {}
+        base_dim_query_names = self._config.get_base_dimension_query_names()
+
+        def default_records(dimension_type):
+            return (
+                self._config.get_base_dimension(dimension_type)
+                .get_records_dataframe()
+                .select("id")
+            )
+
+        for dim_filter in context.model.project.dimension_filters:
+            dim_type = dim_filter.dimension_type
+            df = record_ids.get(dim_type, default_records(dim_type))
+            if dim_filter.query_name == base_dim_query_names[dim_type]:
+                df = dim_filter.apply_filter(df, column="id")
+            else:
+                mapping_records = self._config.get_base_to_supplemental_mapping_records(
+                    dim_filter.query_name
+                )
+                required_ids = (
+                    mapping_records.filter("to_id is not NULL").select("from_id").distinct()
```

Comment:
Doesn't `self._config.get_base_to_supplemental_mapping_records(dim_filter.query_name)` get rid of null to_id from the mapping_records?